interface/src/app/signature/enriched_signatures.js

```diff
@@ -0,0 +1,203 @@
+        response.objects.forEach(function(element) {
+          self.signatures[element.id] = element.name;
+        });
+      },
+      function error(err) {
+        var message = errGen('Failed to get signatures: ', err);
+        $log.error(message);
+        self.statusMessage = message + '. Please try again later.';
+      }
+    ).$promise;
+
+    // Promise that gets the total number of genes:
+    var p2 = Gene.get(
+      {},
+      function success(response) {
+        self.geneNum = response.meta.total_count;
+      },
+      function error(err) {
+        var message = errGen('Failed to get total gene number: ', err);
+        $log.error(message);
+        self.statusMessage = message + '. Please try again later.';
+      }
+    ).$promise;
+
+    var participations;
+    // Promise that gets all participation records that are related to
+    // the genes in URL:
+    var p3 = Participation.get(
+      {'related_genes': $stateParams.genes, 'limit': 0},
+      function success(response) {
+        participations = response.objects;
+      },
+      function error(err) {
+        var message = errGen(
+          'Failed to get signature-gene participations: ', err);
+        $log.error(message);
+        self.statusMessage = message + '. Please try again later.';
+      }
+    ).$promise;
+
+    var pValueSigDigits = 3;
+    // groupedGenes is an object that behaves like a 2-D table of genes.
+    // groupedGenes[participationTypeID][signatureID] is an object whose
+    // keys are gene IDs and values are genes names of those genes that
+    // participate with signatureID and with participationTypeID.
+    var groupedGenes = Object.create(null);
+
+    // Helper function that calculates the enrichment for each signature
+    // that includes at least one gene in the url.
+    var calculateEnrichments = function(typeName, genesBySignatures) {
+      var m = genesInUrl.length;
+      var N = self.geneNum;
+      var pValueArray = [];
+      var matchedGenesBySignature = [];
+      var signatureIDs = Object.keys(genesBySignatures);
+      signatureIDs.forEach(function(sigID) {
+        var genes = genesBySignatures[sigID];
+        var n = Object.keys(genes).length, k = 0;
+        var matchedGenes = [], selectedGene;
+        var pValue;
+        for (var i = 0; i < genesInUrl.length; i++) {
+          selectedGene = genesInUrl[i];
+          if (genes[selectedGene]) {
+            k++;
+            matchedGenes.push(genes[selectedGene]);
+          }
+        }
+        matchedGenesBySignature.push(matchedGenes);
+        // Four parameters are needed in the formula:
+        //     hyperGeometricTest(k, m, n, N)
+        // k: the number of heavy genes related to sigID in genesInUrl;
+        // m: the number of genes in genesInUrl;
+        // n: the total number of heavy genes related to sigID;
+        // N: the total number of genes.
+        pValue = 1 - MathFuncts.hyperGeometricTest(k, m, n, N);
```

Comment:
I need to pull out my stats notes to try and follow this test more carefully. In the meantime, can you tell me how you came up with the inputs here? As I mentioned above, I am not sure the `N` that's going in is what we want. We should double-check the rest of the parameters, too. Ideally, we would have a test case to confirm that the output matches what we expect.

Reply:
I had a conversation with @rzelayafavila about this. The ideal set of genes was the set of genes that _could have actually been returned_. This set contains the genes that were included in the model & that are also in our database. Obtaining that was a time consuming query (and it meant we would need to filter only the query genes to this same set).

IIRC @rzelayafavila checked into this and that number was essentially the same as the number of PA genes in our database. So I think that he decided to use just _the number of genes that exist in the DB for this organism_. This will be serviceable until we go to systems where those differ substantially. I think he was going to file an issue on this and/or add a comment to note the limitations of this approach and why he was selecting it at this time.

Of course, if you want to revisit this and/or you're not concerned about those filters, it'd be possible to do the original/ideal query.

Reply:
Rene's original PR includes the following comment:
```
    // For the moment, the total number of genes in our universe
    // to perform this enrichment analysis, is the total number of
    // Entrez IDs in our database for Pseudomonas aeruginosa.
```
I think this is probably the comments that @cgreene was referring to. (I replaced these lines with a shorter line because it seemed redundant.) I will add the comments back.

Reply:
Yes, exactly, that was the reasoning for getting the total number of genes from the DB for Pseudomonas.

Reply:
Okay, that makes way more sense with the comment then. How can we ensure that the query only returns genes for *pseudomonas*, then? Looking back at `enriched_signatures.js`, line 76, we request *all* genes from the database (without any filter).

I guess you're saying that our database *only* has genes for *pseudomonas*. At some point we plan to add other organisms... how will we remember to fix that line in the future when we begin work on that?

We need an issue, for sure, and we have a milestone called `Before multi-organism`, so the issue should get filed and tagged that way. I think I would take one additional step and add a line of code to our management command for loading genes that will make it halt with an error if we try to load genes for any organism other than *pseudomonas*...

...aha. We can't alter the management command because it is from `django-genes`. The next best thing is a comment in the scripts that call it. I see two places: `adage_server.py`, line 180, and `load_default_pseudomonas_data.sh`, line 40. The comments should make specific mention of the relevant issue.

Reply:
Another approach is to add a unit test that will fail (and halt deployment) if it finds genes from any organism other than *pseudomonas* (`taxonomy_id`=208964). The comments in the test should make specific reference to the issue.